lib/utils/workpool/workpool.go

```diff
@@ -0,0 +1,263 @@
+/*
+Copyright 2020 Gravitational, Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package workpool
+
+import (
+	"context"
+	"sync"
+
+	"go.uber.org/atomic"
+)
+
+// Pool manages a collection of work groups by key and is the primary means
+// by which groups are managed.  Each work group has an adjustable target value
+// which is the number of target leases which should be active for the given
+// group.
+type Pool struct {
+	sync.Mutex
+	leaseIDs *atomic.Uint64
+	groups   map[interface{}]*group
+	grantC   chan Lease
+	ctx      context.Context
```

Comment:
would be helpful to expose the context `Done()` channel here, so the following is possible:

```
select {
   case pool.Done(): 
     pool is done
   case pool.Acquire():
     new lease arrived
```

Right now it's not possible to tell if the pool is stopped while being blocked on the acquire channel